sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -1345,6 +1339,12 @@ public PCollection<KafkaRecord<K, V>> expand(PBegin input) {
       final KafkaIOReadImplementationCompatibilityResult compatibility =
           KafkaIOReadImplementationCompatibility.getCompatibility(this);
 
+      // Reading from Kafka SDF is currently broken, as re-starting the pipeline will cause the
+      // consumer to start from scratch. See https://github.com/apache/beam/issues/21730.
```

Comment:
Please create a new Github issue that more generically covers issues with SDF Kafka and test coverage and reference here.